### PR TITLE
WIP: start adding Transformer based audio encoder

### PR DESCRIPTION
**Pull request by egpbos**
I just started a very rough first implementation, it's not working yet, but I'll just keep updating it in this work-in-progress PR.

### CODE REVIEW

**Comment by egpbos** [comment]
It works now (on my machine and on CI), suggestions / comments very welcome.

**Comment by egpbos** [comment]
Also implemented the optional config parameters scheme I suggested in #5.

File: platalea/encoders.py [hunk 1]
```diff
@@ -140,6 +146,62 @@ def introspect(self, input, length):
...
+        def default_transformer_layer(**config):
+            return nn.TransformerEncoder(nn.TransformerEncoderLayer(**config), num_layers)
+        trafo_layer_type = config.get('trafo_layer_type', default_transformer_layer)
+        self.Transformer = trafo_layer_type(**trafo)
+
+        upsample = config['upsample']
+        if trafo['d_model'] == conv['out_channels']:
+            self.scale_conv_to_trafo = nn.Identity()
+        else:
+            self.scale_conv_to_trafo = nn.Upsample(size=trafo['d_model'], **upsample)
```
Comment (bhigy) [inline platalea/encoders.py]: It would make more sense to use a linear layer than upsampling.

File: platalea/encoders.py [hunk 2]
```diff
@@ -140,6 +146,62 @@ def introspect(self, input, length):
...
+
+        att = config['att']
+        self.att = Attention(**att)
```
Comment (bhigy) [inline platalea/encoders.py]: If we want to be able to use the same encoder for ASR, the attention part should be option (like in `SpeechEncoder`).

**Review by bhigy — COMMENTED** [review]
Based on our recent discussion, I reverted the move of `epochs` from the experiment to the global config, as it doesn't make sense for some scripts (e.g. the utilities). We could create a function somewhere to add to the parser all experiment parameters that are generic like this one.

File: platalea/encoders.py (hunk 1 above)
Comment (egpbos) [reply platalea/encoders.py]: Switched to Linear.

File: platalea/encoders.py (hunk 2 above)
Comment (egpbos) [reply platalea/encoders.py]: Made it optional.

**Comment by egpbos** [comment]
Ok, thanks for the review! Applied your suggestions, merging now so I can train based on this model.

**Pull request merged**